node/api.go

```diff
@@ -61,83 +58,55 @@ func (api *PrivateAdminAPI) AddPeer(url string) (bool, error) {
-	var offeredAPIs []rpc.API
-	if len(apis) > 0 {
-		namespaces := set.New()
-		for _, a := range strings.Split(apis, ",") {
-			namespaces.Add(strings.TrimSpace(a))
-		}
-		for _, api := range api.node.APIs() {
-			if namespaces.Has(api.Namespace) {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	} else { // use by default all public API's
-		for _, api := range api.node.APIs() {
-			if api.Public {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	}
+func (api *PrivateAdminAPI) StartRPC(host string, port int, cors string, apis string) (bool, error) {
+	api.node.lock.Lock()
+	defer api.node.lock.Unlock()
 
-	if address == "" {
-		address = "127.0.0.1"
+	if api.node.httpHandler != nil {
+		return false, fmt.Errorf("HTTP RPC already running on %s", api.node.httpEndpoint)
 	}
-	if port == 0 {
-		port = 8545
+	if err := api.node.startHTTP(fmt.Sprintf("%s:%d", host, port), api.node.rpcAPIs, strings.Split(apis, ","), cors); err != nil {
+		return false, err
 	}
-
-	corsDomains := strings.Split(cors, " ")
-	err := rpc.StartHTTP(address, port, corsDomains, offeredAPIs)
-	return err == nil, err
+	return true, nil
 }
 
 // StopRPC terminates an already running HTTP RPC API endpoint.
 func (api *PrivateAdminAPI) StopRPC() (bool, error) {
-	err := rpc.StopHTTP()
-	return err == nil, err
+	api.node.lock.Lock()
+	defer api.node.lock.Unlock()
+
+	if api.node.httpHandler == nil {
+		return false, fmt.Errorf("HTTP RPC not running")
+	}
+	api.node.stopHTTP()
+	return true, nil
 }
 
 // StartWS starts the websocket RPC API server.
-func (api *PrivateAdminAPI) StartWS(address string, port int, cors string, apis string) (bool, error) {
-	var offeredAPIs []rpc.API
-	if len(apis) > 0 {
-		namespaces := set.New()
-		for _, a := range strings.Split(apis, ",") {
-			namespaces.Add(strings.TrimSpace(a))
-		}
-		for _, api := range api.node.APIs() {
-			if namespaces.Has(api.Namespace) {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	} else {
-		// use by default all public API's
-		for _, api := range api.node.APIs() {
-			if api.Public {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	}
+func (api *PrivateAdminAPI) StartWS(host string, port int, cors string, apis string) (bool, error) {
+	api.node.lock.Lock()
+	defer api.node.lock.Unlock()
 
-	if address == "" {
-		address = "127.0.0.1"
+	if api.node.wsHandler != nil {
+		return false, fmt.Errorf("WebSocker RPC already running on %s", api.node.wsEndpoint)
 	}
-	if port == 0 {
-		port = 8546
+	if err := api.node.startWS(fmt.Sprintf("%s:%d", host, port), api.node.rpcAPIs, strings.Split(apis, ","), cors); err != nil {
+		return false, err
 	}
-
-	corsDomains := strings.Split(cors, " ")
-
-	err := rpc.StartWS(address, port, corsDomains, offeredAPIs)
-	return err == nil, err
+	return true, nil
 }
 
 // StopRPC terminates an already running websocket RPC API endpoint.
 func (api *PrivateAdminAPI) StopWS() (bool, error) {
-	err := rpc.StopWS()
-	return err == nil, err
+	api.node.lock.Lock()
```

Comment:
The reason I needed to lock in the API is because these start/stop methods assume that the lock is already held. This is because they are using during the huge startup procedure, where the lock's already held. Since these are private methods that cannot be accessed outside the `node` package, I think having explicit access to the lock should be fine. The other solution would be adding a set of wrapper methods into the node that just locks/unlocks and calls these. Similarly the node itself never starts/stops the servers on its own or provides an API for for, so it should be fine to error out in this API struct. I don't see too much value in programatically starting/stopping HTTP/WS endpoints on a node, so that's why I didn't provide public methods to do that.